File: internal/domain/tag.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// Tag is a label that can be applied to pages.
type Tag struct {
	ID            uuid.UUID `json:"id"`
	Name          string    `json:"name"`
	Color         string    `json:"color"`
	IsAIGenerated bool      `json:"is_ai_generated"`
	PageCount     int       `json:"page_count,omitempty"`
	CreatedAt     time.Time `json:"created_at"`
}

// CreateTagRequest carries the fields needed to create a new Tag.
type CreateTagRequest struct {
	Name  string `json:"name"`
	Color string `json:"color"`
}

// PageTagAssignment represents associating a tag with a page at a given confidence level.
type PageTagAssignment struct {
	TagID           uuid.UUID `json:"tag_id"`
	ConfidenceScore float64   `json:"confidence_score"`
}

// TagWithScore is a tag augmented with the confidence score for a page assignment.
type TagWithScore struct {
	ID              uuid.UUID `json:"id"`
	Name            string    `json:"name"`
	Color           string    `json:"color"`
	ConfidenceScore float64   `json:"confidence_score"`
}

File: web/lib/models/tag.dart
class Tag {
  final String id;
  final String name;
  final String? color;
  final bool isAiGenerated;
  final int? pageCount;
  final double? confidenceScore;
  final DateTime? createdAt;

  const Tag({
    required this.id,
    required this.name,
    this.color,
    this.isAiGenerated = false,
    this.pageCount,
    this.confidenceScore,
    this.createdAt,
  });

  factory Tag.fromJson(Map<String, dynamic> json) {
    return Tag(
      id: json['id'] as String,
      name: json['name'] as String,
      color: json['color'] as String?,
      isAiGenerated: json['is_ai_generated'] as bool? ?? false,
      pageCount: json['page_count'] as int?,
      confidenceScore: json['confidence_score'] != null
          ? (json['confidence_score'] as num).toDouble()
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'color': color,
      'is_ai_generated': isAiGenerated,
      'page_count': pageCount,
      'confidence_score': confidenceScore,
      'created_at': createdAt?.toIso8601String(),
    };
  }
}
